Simlar/src/SMLRLinphoneHandler.m

```diff
@@ -341,37 +355,92 @@ - (void)destroyLibLinphone
+    NSString *const newOutput = [SMLRLinphoneHandler getRouteOutput:[sharedAudioSession currentRoute]];
 
-    if (_phoneManagerDelegate) {
-        const BOOL oldRouteEnabledSpeaker = [SMLRLinphoneHandler isExternalSpeakerEnabledInRoute:
-                                             notification.userInfo[AVAudioSessionRouteChangePreviousRouteKey]];
-        const BOOL newRouteEnabledSpeaker = [SMLRLinphoneHandler isExternalSpeakerEnabledInRoute:
-                                             [[AVAudioSession sharedInstance] currentRoute]];
+    if (![oldOutput isEqualToString:newOutput]) {
+        SMLRLogI(@"audio session output route changed: %@ -> %@", oldOutput, newOutput);
+    }
 
-        if (oldRouteEnabledSpeaker != newRouteEnabledSpeaker) {
-            SMLRLogI(@"audioSessionRouteChanged: external speaker now: %@", newRouteEnabledSpeaker ? @"enabled" : @"disabled");
-            [_phoneManagerDelegate onExternalSpeakerChanged:newRouteEnabledSpeaker];
+    if (_callStatus.enumValue == SMLRCallStatusEnded) {
+        [self updateAudioOutputType:SMLRAudioOutputTypeNormal];
+    } else {
+        if ([SMLRLinphoneHandler isBlueToothAvailable]) {
+            [self updateAudioOutputType:SMLRAudioOutputTypeBlueToothAvailable];
+        } else if ([AVAudioSessionPortBuiltInSpeaker isEqualToString:newOutput]) {
+            [self updateAudioOutputType:SMLRAudioOutputTypeExternalSpeaker];
+        } else {
+            [self updateAudioOutputType:SMLRAudioOutputTypeNormal];
+        }
+
+        if (![category isEqualToString:[self generateAudioSessionCategory]] || categoryOptions != [self generateAudioSessionCategoryOptions]) {
+            [self updateAudioSession];
         }
     }
 }
 
-+ (BOOL)isExternalSpeakerEnabledInRoute:(AVAudioSessionRouteDescription *const) route
+- (void)updateAudioOutputType:(const SMLRAudioOutputType)type
 {
-    for (AVAudioSessionPortDescription const* desc in [route outputs]) {
-        if ([AVAudioSessionPortBuiltInSpeaker isEqualToString:desc.portType]) {
+    if (_audioOutputType == type) {
+        return;
+    }
+
+    SMLRLogI(@"updating audio output type: %@ -> %@", nameForSMLRAudioOutputType(_audioOutputType), nameForSMLRAudioOutputType(type));
+    self.audioOutputType = type;
+
+    [_phoneManagerDelegate onAudioOutputTypeChanged:type];
+}
+
++ (BOOL)isBlueToothAvailable
+{
+    for(AVAudioSessionPortDescription *const desc in [[AVAudioSession sharedInstance] availableInputs]) {
+        if ([AVAudioSessionPortBluetoothHFP isEqualToString:desc.portType]) {
             return YES;
         }
     }
     return NO;
 }
 
++ (NSString *)getRouteOutput:(AVAudioSessionRouteDescription *const)route
+{
+    const NSUInteger count = [[route outputs] count];
+
+    if (count == 0) {
+        SMLRLogE(@"no output found in audio session: %@", route);
+        return nil;
+    }
+
+    if (count != 1) {
+        SMLRLogW(@"unexpected output count: %lu", (unsigned long) count);
```

Comment:
Do we use a whitespace after cast? Or: `SMLRLogW(@"unexpected output count: %lu", (unsigned long)count);`